feat(instrumentation-winston): add log sending to Logs API

hectorhdzg opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

Fixes #1558 

Winston instrumentation will emit recordLog using Logs API whenever configured on, adding similar configuration options as the ones added for Bunyan in #1713

seemk commented on plugins/node/opentelemetry-instrumentation-winston/README.md:
```diff
@@ -46,15 +40,43 @@ logger.info('foobar');
-### Fields added to Winston metadata
+### Log sending
+
+Winston Logger will automatically send log records to the OpenTelemetry Logs SDK. The OpenTelemetry SDK can be configured to handle those records -- for example, sending them on to an OpenTelemetry collector for log archiving and processing. The example above shows a minimal configuration that emits OpenTelemetry log records to the console for debugging.
```
Is the `--` intentional here?

seemk commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -163,44 +185,97 @@ export class WinstonInstrumentation extends InstrumentationBase {
+        attributes[key] = splat[key];
+      }
+    }
+    const timestamp = Date.now();
```
In the [logs API](https://github.com/open-telemetry/opentelemetry-js/blob/f665499/experimental/packages/api-logs/src/types/LogRecord.ts#L56) the comment states that timestamp should be in nanoseconds.

The spec also specifies [nanoseconds](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-timestamp)

I wonder if the logs API timestamp field should be a `HrTime` instead?

hectorhdzg replied:
Yeah that make sense, updated to use hrTime and convert it to nanoseconds

mhennoch commented on plugins/node/opentelemetry-instrumentation-winston/README.md:
```diff
@@ -18,23 +18,17 @@ npm install --save @opentelemetry/instrumentation-winston
-    // other instrumentations
   ],
 });
 
```
Maybe also call sdk.start() ?

mhennoch commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -39,10 +47,24 @@ import { VERSION } from './version';
+      VERSION,
+      Object.assign({}, DEFAULT_CONFIG, config)
+    );
+    this._logger = logs.getLogger(
```
I wonder if this works with the example in readme.  sdk.start() sets the global log provider but this constructor is ran before it so this._logger will always be noop no?

blumamir commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -219,3 +294,54 @@ function injectRecord(
+};
+
+function getSeverityNumber(level: string): SeverityNumber {
+  let result = npmLevels[level];
+  if (result) {
+    return result;
+  }
+  result = sysLoglevels[level];
+  if (result) {
+    return result;
+  }
+  result = cliLevels[level];
+  if (result) {
+    return result;
+  }
+  // Unknown
+  return SeverityNumber.TRACE;
```
Since `severityNumber` [is optional](https://github.com/open-telemetry/opentelemetry-js/blob/dcf93e85065a2cb2d255278421d39a7b57941b72/experimental/packages/api-logs/src/types/LogRecord.ts#L66) in the API and [in the spec](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/data-model.md#field-severitynumber):

> Description: numerical value of the severity, normalized to values described in this document. This field is optional.

I suggest that this function returns `SeverityNumber | undefined` and when the instrumentation is unable to calculte a known SeverityNumber, it will return `undefined`.

blumamir replied:
nit: you can do it all in one single expression if you like:
```suggestion
  return npmLevels[level] ?? sysLoglevels[level] ?? cliLevels[level];
```

It's a matter of personal style so feel free to use or resolve :)

blumamir commented on plugins/node/opentelemetry-instrumentation-winston/README.md:
```diff
@@ -46,15 +40,43 @@ logger.info('foobar');
+
+Log injection can be disabled with the `disableLogCorrelation: true` option.
+
+### Winston instrumentation options
+
+| Option                  | Type              | Description |
+| ----------------------- | ----------------- | ----------- |
+| `disableLogSending`     | `boolean`         | Whether to disable [log sending](#log-sending). Default `false`. |
+| `disableLogCorrelation` | `boolean`         | Whether to disable [log correlation](#log-correlation). Default `false`. |
+| `logHook`               | `LogHookFunction` | An option hook to inject additional context to a log record after trace-context has been added. This requires `disableLogCorrelation` to be false. |
```
I think it is more useful if the options table is presented before the detailed description about each option. e.g. - move this md section up to be after `Usage`

blumamir commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -138,22 +159,24 @@ export class WinstonInstrumentation extends InstrumentationBase {
-        if (!span) {
-          return original.apply(this, args);
+        if (!config.disableLogCorrelation) {
+          const span = trace.getSpan(context.active());
```
nit: you can use a shortcut for that in the API:
```suggestion
          const span = trace.getActiveSpan();
```

blumamir commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -163,44 +186,96 @@ export class WinstonInstrumentation extends InstrumentationBase {
 
-        args.splice(insertAt, 0, record);
-        instrumentation._callHook(span, record);
+            let msgArguments = 0;
+            for (let i = 1; i < args.length; i++) {
+              if (typeof args[i] === 'string') {
+                msgArguments++;
+              }
+            }
```
nit: You can transform this code to a no-side-effects, shorter version like so:
```suggestion
const msgArguments = args.slice(1).filter(a => typeof a === 'string').length
```

or (if the first argument is always string):
```suggestion
const msgArguments = args.filter(a => typeof a === 'string').length - 1
```

hectorhdzg replied:
I updated this code a little bit, so it doesn't apply anymore, let me know if some other code can be transformed as well.

blumamir commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -163,44 +186,96 @@ export class WinstonInstrumentation extends InstrumentationBase {
+                logRecord['msg'] = args[1];
+              } else {
+                // Handle string interpolation
+                const values = args.slice(2, msgArguments + 1);
+                logRecord['msg'] = util.format(args[1], ...values);
```
I didn't run it myself, but it seems [winston describes that this behavior is not always applied](https://github.com/winstonjs/winston#string-interpolation), and if it does, the arguments might be non-string:

> The log method provides the string interpolation using [util.format](https://nodejs.org/dist/latest/docs/api/util.html#util_util_format_format_args). It must be enabled using format.splat().

and: 

```js
logger.log('info', 'test message %d', 123);
```

hectorhdzg replied:
Updated the code, util.format support multiple types so passing them as they are now, FYI this code is only applicable to Winston 2 https://github.com/winstonjs/winston/tree/2.x, string interpolation in newer version is handled before write is called.

blumamir commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -163,44 +186,96 @@ export class WinstonInstrumentation extends InstrumentationBase {
-            ? args.length - 1
-            : args.length;
+            // Get meta if present
+            for (let i = args.length - 1; i >= 0; i--) {
+              if (typeof args[i] === 'object') {
+                logRecord['meta'] = args[i];
+                break;
+              }
+            }
```
you can do it in one go like so:
```js
logRecord['meta'] = args.findLast(a => typeof a === 'object')
```

hectorhdzg replied:
I would need to cast args to an array to use that, I usually don't risk casting for args out of my control, also added some other logic to calculate the index to quickly determine the string interpolation parameters.

blumamir commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -163,44 +186,96 @@ export class WinstonInstrumentation extends InstrumentationBase {
+          const span = trace.getSpan(context.active());
+          if (span) {
+            const spanContext = span.spanContext();
+            if (isSpanContextValid(spanContext)) {
```
also here, previous to this PR, when `!isSpanContextValid(spanContext)` we would `return original.apply(this, args);` and skip calling the hook and injecting anything. now we will do it even for invalid contexts

blumamir commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -163,44 +186,96 @@ export class WinstonInstrumentation extends InstrumentationBase {
-            injectRecord(spanContext, record);
-            instrumentation._callHook(span, record);
-            return original.apply(this, args);
+              args.splice(insertAt, 0, record);
```
we are manipulating the `args` here, and then use the updated value below for the log bridge. Is it ok?

hectorhdzg replied:
Well the span information would be duplicated in this case I guess if Logger is also adding correlation information into the LogRecords

blumamir reviewed (commented):
Thanks for working on this 🥇 

Overall, LGTM. Added a few minor comments.

hectorhdzg commented on plugins/node/opentelemetry-instrumentation-winston/package.json:
```diff
@@ -41,7 +41,8 @@
   "peerDependencies": {
-    "@opentelemetry/api": "^1.3.0"
+    "@opentelemetry/api": "^1.3.0",
+    "@opentelemetry/api-logs": "^0.45.1"
```
Well I believe we should update Bunyan as well, I can create a different PR so people are aware specific changes are happening for that instrumentation, some issues in the readme that were brough up by @mhennoch  should apply to bunyan as well

martinkuba commented:
I think it would be better to implement this as a Winston transport rather than an instrumentation, similar to how it has been implemented for [Bunyan as a stream](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/plugins/node/opentelemetry-instrumentation-bunyan/src/OpenTelemetryBunyanStream.ts). The [Logs Bridge API](https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/logs/bridge-api.md) was intended for log appenders, not instrumentations.

hectorhdzg commented:
@martinkuba my initial approach was using Winston transports, and the problem I found was related to having a direct dependency to winston 3x and 2x to make it work, transports require to extend a specific class in winston is not just to follow some interface, so that would mean people installing @opentelemetry-instrumentation-winston will get winston 3x and winston 2x installed as well, I believe this is undesirable and that is why I decided to use existing patch in write/log methods instead. Related to log appenders or instrumentations, there was some discussion about this, we can create another new appender package just for Winston, Bunyan and others, but I believe the agreement was to use the instrumentation instead, I don't have a problem creating a new appender project here, but is just some maintenance overhead in my opinion and possible causing users confusion.

martinkuba commented:
@hectorhdzg That makes sense. I just wanted to highlight that the Logs API was not intended to be used in instrumentations (per my interpretation of the spec). Maybe it's fine to approach it this way, but I am also thinking in terms of allowing (or encouraging) instrumentations to generate logs. The PR linked below is related.
https://github.com/open-telemetry/opentelemetry-js/pull/4314

I see that Java has different packages for different versions of Log4j appenders
https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/log4j

hectorhdzg commented:
@martinkuba I understand appender term makes sense as is what the specification mentions, I created a discussion issue to talk about this in a single place, as this conversation has been happening in several places. I don't have strong opinions about one or the other, I just want to allow this functionality and move this forward.
https://github.com/open-telemetry/opentelemetry-js/issues/4355

trentm commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -163,40 +181,93 @@ export class WinstonInstrumentation extends InstrumentationBase {
+              newTransports = [originalTransports];
+            }
+            newTransports.push(openTelemetryTransport);
+            originalTransports = newTransports;
```
This misses adding the OpenTelemetryTransportv3() if `transports` is not part of the Winston logger options:

```sh
% cat telemetry.js
// Setup telemetry for tracing and Winston logging.
const { NodeSDK, tracing, logs, api } = require('@opentelemetry/sdk-node');
// api.diag.setLogger(new api.DiagConsoleLogger(), api.DiagLogLevel.DEBUG);
const { WinstonInstrumentation } = require('./');
const sdk = new NodeSDK({
  serviceName: 'winston-example',
  spanProcessor: new tracing.SimpleSpanProcessor(new tracing.ConsoleSpanExporter()),
  logRecordProcessor: new logs.SimpleLogRecordProcessor(new logs.ConsoleLogRecordExporter()),
  instrumentations: [
    new WinstonInstrumentation(),
  ]
})
process.on("SIGTERM", () => {
  sdk.shutdown().finally(() => process.exit(0));
});
sdk.start();

% cat no-transports.js
const winston = require('winston');
const logger = winston.createLogger({
  level: 'info',
  // transports: [
  //   new winston.transports.Console(),
  // ]
});
logger.info('hi there')

% node -r ./telemetry.js no-transports.js
[winston] Attempt to write logs with no transports {"message":"hi there","level":"info"}
```

Granted this is an edge case.

A fix would be:

```diff
diff --git a/plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts b/plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts
index f06225d1..3d8cbbd1 100644
--- a/plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts
+++ b/plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts
@@ -229,7 +229,7 @@ export class WinstonInstrumentation extends InstrumentationBase {
         const config = instrumentation.getConfig();
         if (!config.disableLogSending) {
           if (args && args.length > 0) {
-            let originalTransports = args[0].transports;
+            const originalTransports = args[0].transports;
             let newTransports = Array.isArray(originalTransports)
               ? originalTransports
               : [];
@@ -238,7 +238,7 @@ export class WinstonInstrumentation extends InstrumentationBase {
               newTransports = [originalTransports];
             }
             newTransports.push(openTelemetryTransport);
-            originalTransports = newTransports;
+            args[0].transports = newTransports;
           }
         }
         return original.apply(this, args);
```

Then:

```sh
% node -r ./telemetry.js no-transports.js
{
  timestamp: 1705507494522000,
  traceId: undefined,
  spanId: undefined,
  traceFlags: undefined,
  severityText: 'info',
  severityNumber: 9,
  body: 'hi there',
  attributes: {}
}
```

That *does* mean the user no longer gets the Winston warning. However that was already going to be the case with an empty transports array:

```js
const winston = require('winston');
const logger = winston.createLogger({
  level: 'info',
  transports: []
});
logger.info('hi there')
```

trentm commented on plugins/node/opentelemetry-instrumentation-winston/src/OpenTelemetryTransportv3.ts:
```diff
@@ -0,0 +1,44 @@
+  public override log(info: any, next: () => void) {
+    try {
+      emitLogRecord(info, this._logger);
+      this.emit('logged', info);
```
1. The usage at https://github.com/winstonjs/winston-transport#usage suggests emitting `logged` should be done async (e.g. in `setImmediate`) though it doesn't document why. Looking briefly at the core transports in winston@3, they all seem to do so. Perhaps this should as well?
2. Also, the `this.emit('logged', ...)` should move *out* of the try/catch, otherwise a thrown exception in a handler for the 'logged' event will get caught here.

trentm commented on plugins/node/opentelemetry-instrumentation-winston/test/winston.test.ts:
```diff
@@ -34,31 +40,59 @@ const tracer = provider.getTracer('default');
+        // format: winston.format.combine(
+        //   winston.format.splat(),
+        //   winston.format.simple()
+        // )
```
nit: Debugging code to remove?

trentm commented on plugins/node/opentelemetry-instrumentation-winston/package.json:
```diff
@@ -58,12 +59,14 @@
+    "@opentelemetry/api-logs": "^0.45.1",
+    "@opentelemetry/core": "1.18.1",
+    "@opentelemetry/instrumentation": "^0.46.0",
+    "winston": "3.3.3",
```
Most (all?) the winston@3 transports I see have a dependency just on `winston-transport`. Could this be changed to only depend on it, and not take a runtime dependency on `winston` itself?

Flarna replied:
The guild [here](https://github.com/open-telemetry/opentelemetry-js-contrib/blob/main/GUIDELINES.md#instrumented-package-dependency) explains why instrumentations should not depend on the instrumented library.

I think the instrumentation has to find the winston-transport instances (there might be several of different versions) via the hooking mechanism used for patching. If this isn't possible some other mechanism should be used to avoid depending on the instrumented library.

trentm replied:
@Flarna If this PR changes to dep on `winston-transport` would that satisfy the guidelines? This is instrumenting `winston`, not `winston-transport`.

One possible use-case of this package -- as described at https://github.com/open-telemetry/opentelemetry-js-contrib/blob/4c6e81347bb1dbc0c87a0de484a48354cef46cb2/plugins/node/opentelemetry-instrumentation-winston/README.md#using-opentelemetrywinstontransport-without-instrumentation -- will be for a user to directly use the `OpenTelemetryTransportv3` class to configure their Winston logger. In that case there isn't necessary an active Winston instrumentation. In other words, I don't think it is possible to rely on being able to get `winston-transport` via the hooking mechanism.

Flarna replied:
As long as `@opentelemetry/instrumentation-winston`  is part of `@opentelemetry/auto-instrumentations-node` I would avoid such dependencies because
* it adds winston (incl deps) even in projects where winston is not used
* it has the risk of causing double installs because version ranges do not match
* might cause issuse with types (see #932)

Maybe we should remove winston instrumentation from `@opentelemetry/auto-instrumentations-node`. OptIn will be likely done by winston users only.
Or split packages into the "classic" instrumentation (part of auto package) and the transport extension (opt-in).

trentm commented on plugins/node/opentelemetry-instrumentation-winston/package.json:
```diff
@@ -58,12 +59,14 @@
+    "@opentelemetry/core": "1.18.1",
+    "@opentelemetry/instrumentation": "^0.46.0",
+    "winston": "3.3.3",
+    "winston2": "npm:winston@2.4.7"
```
Having a runtime dependency on winston@2 would be unfortunate. Would it be possible to just have a peer dep instead? Most of the winston@2 era transports that I looked at do *not* have a direct dependency on `winston` listed in their package.json. (FWIW the don't list `winston` in `peerDependencies` either, mostly I think because winston@2 predated `peerDependencies` existing or being well-known in npm.)

This instrumentation having a runtime dep on `winston@2` means carrying along its transitive deps, and winston@2's last release was 2022-11-15. Those deps aren't *that* big (mostly it is ~4MB from lodash); and the deps don't currently have an security warnings.

If it is hard/impossible to make a winston@2 transport subclass without a direct dep on winston@2, a possible consideration is to only add log sending support for winston@3.

hectorhdzg replied:
Ok, I checked a couple of winston custom transports and basically they have winston as a peer dependency, and for winston <2 they basically only have a dev dependency, so they will fail if winston is not also installed in the app. Easy fix here is to add a peer dependency but like @Flarna mentioned in other comment this would cause several issues and does not align with current guidelines, I can explore suggested approach by @seemk but I can immediately see utils.inherits is a legacy API and discouraged in the official docs, so not sure how to proceed.

seemk replied:
> Ok, I checked a couple of winston custom transports and basically they have winston as a peer dependency, and for winston <2 they basically only have a dev dependency, so they will fail if winston is not also installed in the app. Easy fix here is to add a peer dependency but like @Flarna mentioned in other comment this would cause several issues and does not align with current guidelines, I can explore suggested approach by @seemk but I can immediately see utils.inherits is a legacy API and discouraged in the official docs, so not sure how to proceed.

While it is legacy in Node.js, the [implementation is quite simple](https://github.com/nodejs/node/blob/v21.6.1/lib/util.js#L241-L260), i.e. `Object.defineProperty` and `Object.setPrototypeOf`. Just an idea :slightly_smiling_face:

trentm replied:
I think this is resolved. IIRC when I added this comment the "winston2" dep was in "dependencies" rather than in "devDependencies" where it is now.

trentm reviewed (changes requested):
This looks great. Thanks for working on this, Hector.

pichlermarc reviewed (changes requested):
This looks very good already! :slightly_smiling_face: Looks to me the only thing that needs to be figured out is how we get the winston base code to our users. 

I've added my ideas how we *could* address the situation to the comments, as the current setup will likely cause typescript errors for users that don't have winston installed (couldn't verify yet as it does not compile for me right now, but I've seen similar setups previously that had this effect).

hectorhdzg commented on plugins/node/opentelemetry-instrumentation-winston/package.json:
```diff
@@ -59,11 +60,13 @@
     "typescript": "4.4.4",
-    "winston": "3.3.3",
+    "winston-transport": "4.6.0",
     "winston2": "npm:winston@2.4.7"
```
Thanks @pichlermarc, sounds good, will keep moving forward only allowing winston 3 for now for auto generation of logs, yeah this was an actual dependency not a dev dependency and latest changes are in a weird state as I was updating code addressing comments and testing with dev dependencies at the same time, like you mentioned as a dev dependency will throw an error when used if winston 2 is not present

hectorhdzg commented on plugins/node/opentelemetry-instrumentation-winston/test/winston.test.ts:
```diff
@@ -34,31 +40,59 @@ const tracer = provider.getTracer('default');
             stream,
           }),
         ],
+        // format: winston.format.combine(
+        //   winston.format.splat(),
+        //   winston.format.simple()
+        // )
```
I used to have this code before I added the Transports, supporting string interpolation in the actual write/log method, not needed anymore, forgot to push changes to remove it.

hectorhdzg commented:
@trentm, @pichlermarc, @seemk, @Flarna I updated the code according to @pichlermarc suggestions, created a new package for Winston OpenTelemetry Transport, only version > 3 are currently supported as it rely on winston-transport not available in winston 1 and 2. Added code to dynamically load the package and log an error if not available but it will not throw any error if not there,  customer will need to install 2 packages to have auto generation of logs in Winston now. We can follow up with winston 2 support in the future, PR is quite long now because of 2 different projects being involved, I can split this in 2 PRs if preferred, let me know.

Flarna commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -163,40 +177,74 @@ export class WinstonInstrumentation extends InstrumentationBase {
+            const spanContext = span.spanContext();
+            if (isSpanContextValid(spanContext)) {
+              record = injectRecord(spanContext, record);
+              instrumentation._callHook(span, record);
```
There seems to be a bit of code duplication with `_getPatchedWrite`. Maybe it could be extracted into a helper?

Flarna commented on plugins/node/opentelemetry-instrumentation-winston/src/instrumentation.ts:
```diff
@@ -163,40 +177,74 @@ export class WinstonInstrumentation extends InstrumentationBase {
+              args[0].transports = newTransports;
+            }
+            catch (err) {
+              instrumentation._diag.error('OpenTelemetry Winston transport is not available', err);
```
Not sure if an error log is a good choice here.
The default is `disableLogSending: false` but on default `@opentelemetry/winston-transport` is not installed.
So on default an error log will be created to my understanding.

Maybe change `disableLogSending` to `enableLogSending` (with default false)?
Or use 3 states:
* `enabled` - error log if `@opentelemetry/winston-transport` missing
* `disabled` - never try to use `@opentelemetry/winston-transport`
* `auto`- try to load `@opentelemetry/winston-transport` and use it if present but ignore if not (maybe info/debug log)

hectorhdzg replied:
My only problem changing disableLogSending to enableLogSending is that it will not align with Bunyan instrumentation anymore, changing default is also problematic, I would expect users want this feature to be on by default, maybe changing the error to a info kind of message would be enough. I would argue this diagnostic error or info message is actually useful for customers so they understand something is missing.

Flarna replied:
I'm not sure if everyone really uses an OTLP endpoint for logs. There are existing tools which do log collection even for services which are not instrumented by OpenTelemetry so why should these users switch to direct sending and potentially get duplicate data?

Anyhow, I agree that we should stay consistent. So either establish a new standard and adapt bunyan instrumentation or define that bunyan is the standard and apply it to others.

Personally I would be even in favor that this is synced across techs but I doubt we will get this into some sort of generic spec in a reasonable timeframe.

hectorhdzg replied:
Ok, changed to enableLogSending with default false, it would be good to change Bunyan to align with this in the future

Flarna replied:
Thought once more about this and I think for the first step it's likely better to stay in sync with bunyan and go for enabled on default.
I forgot that the actual decision of exporting logs is done by configuring SDK and exporter,

Sorry for the back and forth.

Flarna commented on plugins/node/opentelemetry-instrumentation-winston/src/internal-types.ts:
```diff
@@ -14,16 +14,22 @@
-import type { Logger as Winston3Logger } from 'winston';
+import type {
+  Logger as Winston3Logger,
+  LoggerOptions as Winstons3LoggerOptions,
```
is this still needed?

Flarna commented on plugins/node/opentelemetry-winston-transport/src/index.ts:
```diff
@@ -0,0 +1,17 @@
+ * limitations under the License.
+ */
+
+export { OpenTelemetryTransport } from './OpenTelemetryTransport';
```
From package naming and discussion I have the feeling that we might add a transport for winston v2 here.
Should we include the version in the export here to start extensible naming already now?

Flarna commented on plugins/node/opentelemetry-winston-transport/src/OpenTelemetryTransport.ts:
```diff
@@ -0,0 +1,43 @@
+
+  constructor(options?: Transport.TransportStreamOptions) {
+    super(options);
+    this._logger = logs.getLogger('@opentelemetry/winston', VERSION);
```
to my understanding the logger returned here has `includeTraceContext` set to `true`.
Does this fit well to the winston instrumentation which adds tracecontext also into the log line itself?
Isn't this double enrichment?

hectorhdzg replied:
Should we receive the Logger created in the instrumentation here then instead? and create one only if none was provided?

Flarna replied:
My point was not that much about having one or two instances of Logger. It's more about adding traceId to the log line itself and additionally adding it into the export. Isn't this doing the same thing twice?

hectorhdzg replied:
Logger will add traceId of current context or use specific one if part of the actual LogRecord, there should not be any problem here that I can see. 
https://github.com/open-telemetry/opentelemetry-js/blob/main/experimental/packages/sdk-logs/src/Logger.ts#L31
includeTraceContext was removed last year
https://github.com/open-telemetry/opentelemetry-js/pull/3817

Flarna replied:
ah ok, this was removed...
Seems removal was only partially as I found at least [this](https://github.com/open-telemetry/opentelemetry-js/blob/f86251d40fbf615be87319c8a1f5643afb820076/experimental/packages/api-logs/src/types/LoggerOptions.ts#L33) and [this](https://github.com/open-telemetry/opentelemetry-js/blob/f86251d40fbf615be87319c8a1f5643afb820076/experimental/packages/sdk-logs/src/config.ts#L33).

hectorhdzg replied:
Yeah, I can create a PR later to deprecate the config value and clean it up so we don't forget to take it out once Logs API is stable

Flarna commented on plugins/node/opentelemetry-winston-transport/README.md:
```diff
@@ -0,0 +1,68 @@
+## Usage
+
+This package exports the Winston transport class that is used to send records to the
+OpenTelemetry Logs SDK. It can be used directly when configuring a Winston logger
```
Doesn't the setup described here result in double instantiation of the transport? Once by user and once by the instrumentation?

pichlermarc replied:
I agree, it would probably be good to link the docs of `@opentelemetry/instrumentation-winston` to let people know that they don't need to do both to avoid double instantiation.

hectorhdzg replied:
Added some extra details about @opentelemetry/instrumentation-winston

pichlermarc commented on plugins/node/opentelemetry-winston-transport/src/OpenTelemetryTransport.ts:
```diff
@@ -0,0 +1,43 @@
+/*
```
I think it would be great to move this package to the `packages/winston-transport` directory (dropping the `opentelemetry` prefix) instead. Moving this transport package to `packages` would be preferred as it's not necessarily a plugin (what instrumentations were called back in the day).

pichlermarc commented on plugins/node/opentelemetry-winston-transport/CHANGELOG.md:
```diff
@@ -0,0 +1,7 @@
+# Changelog
+
+## [0.1.0]() ()
+
+### Features
+
+* Add Winston OpenTelemetry Transport ([#](https://github.com/open-telemetry/opentelemetry-js-contrib/pull/))
```
The changelog can be empty on a new package, it'll be filled automatically on the first release.

pichlermarc reviewed (commented):
Thanks for making these changes, moving closer to getting this merged :slightly_smiling_face: 

Mostly questions about naming and docs nits from my side. I ran into similar questions as Flarna has already pointed out. 

The transport looks good already, I think pulling it out into a separate PR could speed up the review process and get something ready for manual setup. If you prefer to keep it in one PR that's also fine. I think the diff is still very manageable. :slightly_smiling_face:

pichlermarc commented on release-please-config.json:
```diff
@@ -56,6 +56,7 @@
     "plugins/node/opentelemetry-instrumentation-restify": {},
     "plugins/node/opentelemetry-instrumentation-router": {},
     "plugins/node/opentelemetry-instrumentation-winston": {},
+    "plugins/node/opentelemetry-winston-transport": {},
```
there is also a second file `.release-please-manifest.json` that needs to be adapted. 
doing this will make the `release-please-validate` workflow pass.

hectorhdzg commented:
Thanks @Flarna and @pichlermarc for taking another look at this PR, hopefully we are getting close to merge this one, I moved the winston transport project to packages folder and try to address your feedback, there are some open discussions that hopefully we can resolve soon.

hectorhdzg commented:
@pichlermarc @Flarna @trentm this one is ready for another round, all feedback should be addressed now.

Flarna commented on plugins/node/opentelemetry-instrumentation-winston/README.md:
```diff
@@ -19,47 +19,93 @@ npm install --save @opentelemetry/instrumentation-winston
+
+### Log sending
+
+Winston Logger will automatically send log records to the OpenTelemetry Logs SDK if not explicitly disabled in config and @opentelemetry/winston-transport npm package is installed in the project. The OpenTelemetry SDK can be configured to handle those records, for example, sending them on to an OpenTelemetry collector for log archiving and processing. The example above shows a minimal configurati ...
```
That's a bit out of sync now, also below `disableLogSending` is mentioned.

But as mentioned in another comment going for enabled on default is also fine for me.

hectorhdzg commented on plugins/node/opentelemetry-instrumentation-winston/package.json:
```diff
@@ -63,7 +65,10 @@
   "dependencies": {
-    "@opentelemetry/instrumentation": "^0.48.0"
+    "@opentelemetry/api-logs": "^0.48.0",
+    "@opentelemetry/core": "1.21.0",
```
So many iterations in this PR I forgot to remove this one, not needed anymore

trentm commented on plugins/node/opentelemetry-instrumentation-winston/README.md:
```diff
@@ -19,47 +19,93 @@ npm install --save @opentelemetry/instrumentation-winston
+const tracerProvider = new NodeTracerProvider();
+tracerProvider.register();
+
+// To start a logger, you first need to initialize the Logger provider.
```
```suggestion
// To start a logger, you first need to initialize the Logger provider.
```

Perhaps something like `// For "log sending" a global LoggerProvider must be registered with the OTel API.`

trentm commented on plugins/node/opentelemetry-instrumentation-winston/README.md:
```diff
@@ -19,47 +19,93 @@ npm install --save @opentelemetry/instrumentation-winston
+
+| Option                  | Type              | Description |
+| ----------------------- | ----------------- | ----------- |
+| `enableLogSending`      | `boolean`         | Whether to enable [log sending](#log-sending). Default `false`. |
```
See @Flarna's latest comment on https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1837#discussion_r1474080114
Perhaps we are back to `disableLogSending`?

hectorhdzg replied:
Well then original concern is back again right? we will log an error in our diag Api if transport package is not present, @Flarna thoughts?

hectorhdzg replied:
I see @Flarna comment now, to much stuff happening here :) I can update to disableLogSending defaulting to false same as Bunyan, if everyone is ok with that, the error will logged if transport is not present though

Flarna replied:
Feel free to go for enabled on default to stay consistent with bunyan.
Regardind log: In my opinion info or maybe warning log level would be better but error is also acceptable for me.

trentm replied:
FWIW, I would be fine having it be off by default. (I'd be fine either way.) The OTel docs on logging and the OTel Java SDK config docs make it clear that it is a typical user's choice on whether they want to (a) log to stdout/file with added trace_id et al fields and have a collector ingest them *or* (b) have the SDK directly send logs to some OTLP endpoint.  So that argues for not having both facilities on by default.

trentm commented on plugins/node/opentelemetry-instrumentation-winston/README.md:
````diff
@@ -19,47 +19,93 @@ npm install --save @opentelemetry/instrumentation-winston
+```
+
+Log injection can be disabled with the `disableLogCorrelation: true` option.
+|
````
```suggestion
```

trentm commented on packages/winston-transport/.tav.yml:
I don't think a test-all-versions (TAV) test run is necessary for this package. It is for the instrumentations that work with various versions of the modules that they instrument. This package has a dep on `winston-transport` and really can assume that it stays compatible. If a new `winston-transport` 4.x breaks this, then we just bump the dep range here.

trentm commented on packages/winston-transport/package.json:
```diff
@@ -0,0 +1,59 @@
+  "repository": "open-telemetry/opentelemetry-js-contrib",
+  "scripts": {
+    "test": "nyc ts-mocha -p tsconfig.json 'test/**/*.test.ts'",
+    "test-all-versions": "tav",
```
See comment above. I think the "test-all-versions" script can be dropped here.

trentm reviewed (commented):
Thanks for all the work on this.

In addition to the comments below, I'd suggest this patch do that `test-all-versions` is testing some recent releases of Winston:

```diff
diff --git a/plugins/node/opentelemetry-instrumentation-winston/.tav.yml b/plugins/node/opentelemetry-instrumentation-winston/.tav.yml
index 69b59a03..50007fbc 100644
--- a/plugins/node/opentelemetry-instrumentation-winston/.tav.yml
+++ b/plugins/node/opentelemetry-instrumentation-winston/.tav.yml
@@ -1,5 +1,5 @@
 winston:
-    # a sample from supported versions
-  - versions: "3.3.3 || 3.2.1 || 2.4.5 || 2.4.4 || 2.1.1"
+  # A selection of versions of Winston 2.x and 3.x
+  - versions: "2.0.0 || 2.4.0 || 2.4.7 || 3.3.1 || 3.6.0 || 3.10.0 || ^3.11.0"
     commands: npm run test
```

trentm commented:
Here is a small test where it looks like the OTel LogRecord `traceId` and `spanId` values are not getting set correctly:

```
% pwd
/Users/trentm/tm/opentelemetry-js-contrib5/plugins/node/opentelemetry-instrumentation-winston

% cat foo.js
const otel = require('@opentelemetry/api');
const winston = require('winston');

const logger = winston.createLogger({
  level: 'info',
  transports: [
    new winston.transports.Console(),
  ]
});

logger.info('hi');

const tracer = otel.trace.getTracer('example');
tracer.startActiveSpan('manual-span', span => {
  logger.info('in a span', {foo: 'bar'});
  span.end();
});

% cat telemetry.js
// Setup telemetry for tracing and Winston logging.
const { NodeSDK, tracing, logs, api } = require('@opentelemetry/sdk-node');
// api.diag.setLogger(new api.DiagConsoleLogger(), api.DiagLogLevel.DEBUG);
const { WinstonInstrumentation } = require('./');
const sdk = new NodeSDK({
  serviceName: 'winston-example',
  spanProcessor: new tracing.SimpleSpanProcessor(new tracing.ConsoleSpanExporter()),
  logRecordProcessor: new logs.SimpleLogRecordProcessor(new logs.ConsoleLogRecordExporter()),
  instrumentations: [
    new WinstonInstrumentation({
      enableLogSending: true
    }),
  ]
})
process.on("SIGTERM", () => {
  sdk.shutdown().finally(() => process.exit(0));
});
sdk.start();

% node -r ./telemetry.js foo.js
{"level":"info","message":"hi"}
{"foo":"bar","level":"info","message":"in a span","span_id":"6d5b98c69a377354","trace_flags":"01","trace_id":"99820eddd6eeebaa99d46badf4f37b77"}
{
  timestamp: 1707528570304000,
  traceId: undefined,
  spanId: undefined,
  traceFlags: undefined,
  severityText: 'info',
  severityNumber: 9,
  body: 'hi',
  attributes: {}
}
{
  traceId: '99820eddd6eeebaa99d46badf4f37b77',
  parentId: undefined,
  traceState: undefined,
  name: 'manual-span',
  id: '6d5b98c69a377354',
  kind: 0,
  timestamp: 1707528570306000,
  duration: 1259.054,
  attributes: {},
  status: { code: 0 },
  events: [],
  links: []
}
{
  timestamp: 1707528570311000,
  traceId: undefined,
  spanId: undefined,
  traceFlags: undefined,
  severityText: 'info',
  severityNumber: 9,
  body: 'in a span',
  attributes: {
    foo: 'bar',
    trace_id: '99820eddd6eeebaa99d46badf4f37b77',
    span_id: '6d5b98c69a377354',
    trace_flags: '01'
  }
}
```

Two issues with that last LogRecord:

1. `traceId`, `spanId`, and `traceFlags` are not set to the span context.
2. `attributes` has the trace-context fields, and I think records sent to the Logs SDK should not have those. At least for the Bunyan instrumentation I explicitly masked those out: https://github.com/open-telemetry/opentelemetry-js-contrib/blob/39c34df61f0e60a93fc787dcfceca9782341b837/plugins/node/opentelemetry-instrumentation-bunyan/src/OpenTelemetryBunyanStream.ts#L130-L143


A weird thing here is that if I remove that first `log.info('hi')` -- which I would think would be unrelated -- then the "in a span" LogRecord *does* have the span context fields:

```
% cat foo.js
const otel = require('@opentelemetry/api');
const winston = require('winston');

const logger = winston.createLogger({
  level: 'info',
  transports: [
    new winston.transports.Console(),
  ]
});

// logger.info('hi');

const tracer = otel.trace.getTracer('example');
tracer.startActiveSpan('manual-span', span => {
  logger.info('in a span', {foo: 'bar'});
  span.end();
});

% node -r ./telemetry.js foo.js
{"foo":"bar","level":"info","message":"in a span","span_id":"44b045902e82a7cb","trace_flags":"01","trace_id":"d091897572a459199de5867f13e16ff9"}
{
  timestamp: 1707528800006000,
  traceId: 'd091897572a459199de5867f13e16ff9',
  spanId: '44b045902e82a7cb',
  traceFlags: 1,
  severityText: 'info',
  severityNumber: 9,
  body: 'in a span',
  attributes: {
    foo: 'bar',
    trace_id: 'd091897572a459199de5867f13e16ff9',
    span_id: '44b045902e82a7cb',
    trace_flags: '01'
  }
}
{
  traceId: 'd091897572a459199de5867f13e16ff9',
  parentId: undefined,
  traceState: undefined,
  name: 'manual-span',
  id: '44b045902e82a7cb',
  kind: 0,
  timestamp: 1707528800003000,
  duration: 3241.446,
  attributes: {},
  status: { code: 0 },
  events: [],
  links: []
}
```

Flarna commented:
As far as I remember transports in winston are not sync always. They are even batched in some scenarios like high load or transport chain setup,... because internal streams are used.

So the traceId/spanId added by transport is not that one from log writing in user app - it's that one from transport.
Not sure what the log spec wants here as both might have a valid usecase.

If the traceId/spanId from actual log statement is needed the enrichment by log sdk using getCurrentSpan can't work as it is too late.

pichlermarc reviewed (approved):
Looks like all comments are now addressed. Thanks for sticking with us through the long review process @hectorhdzg.

@trentm I see there's still a changes requested review from your side - would you mind having another look to check if your comments have been addressed? :slightly_smiling_face:

hectorhdzg commented:
@trentm let me know if there is any concern pending

trentm commented on package-lock.json:
```diff
@@ -38646,7 +38476,34 @@
-        "@opentelemetry/api": "^1.3.0"
+        "@opentelemetry/api": "^1.3.0",
+        "@opentelemetry/api-logs": "^0.49.1"
+      }
+    },
+    "plugins/node/opentelemetry-winston-transport": {
+      "name": "@opentelemetry/winston-transport",
+      "version": "0.0.1",
+      "extraneous": true,
+      "license": "Apache-2.0",
+      "dependencies": {
+        "@opentelemetry/api-logs": "^0.48.0",
+        "winston-transport": "4.*"
+      },
+      "devDependencies": {
+        "@types/mocha": "7.0.2",
+        "@types/node": "18.6.5",
+        "@types/sinon": "10.0.18",
+        "@types/triple-beam": "^1.3.2",
+        "mocha": "7.2.0",
+        "nyc": "15.1.0",
+        "rimraf": "5.0.5",
+        "sinon": "15.2.0",
+        "test-all-versions": "6.0.0",
+        "ts-mocha": "10.0.0",
+        "typescript": "4.4.4"
+      },
+      "engines": {
+        "node": ">=14"
```
```suggestion
```

This is a vestigial entry from when this package was in the plugins/node/... dir.
Doing regular `npm install` doesn't clean it up.

hectorhdzg replied:
Wow, how do you catch this kind of things?, I was expecting npm install to always take care of issues like this one, manually removed now.

trentm replied:
> Wow, how do you catch this kind of things?

With all the npm update troubles we've been having (see https://github.com/open-telemetry/opentelemetry-js-contrib/issues/1917 and others), I've gotten into the habit of reading/skimming package-lock diffs. 

> I was expecting npm install to always take care of issues like this

Indeed, that would be nice.  I'm not enamoured with `npm` these days.

trentm commented on packages/winston-transport/package.json:
```diff
@@ -0,0 +1,58 @@
+    "nyc": "15.1.0",
+    "rimraf": "5.0.5",
+    "sinon": "15.2.0",
+    "test-all-versions": "6.0.0",
```
```suggestion
```

nit: This dep isn't used.
If you take this change, you should re-run `npm install` at the root to update the package-lock.json file.

trentm commented on packages/winston-transport/src/utils.ts:
```diff
@@ -0,0 +1,75 @@
+  record: Record<string, any>,
+  logger: Logger
+): void {
+  const { message, msg, level, meta, ...splat } = record;
```
re-up on the `msg` field here.

hectorhdzg replied:
Thanks for bringing this up, this comment got completely lost for my for some reason, msg and meta are winston 2 fields, we should not have this anymore in v3 transport

trentm commented on packages/winston-transport/src/utils.ts:
```diff
@@ -0,0 +1,75 @@
+  logger: Logger
+): void {
+  const { message, msg, level, meta, ...splat } = record;
+  const attributes = Object.assign(meta ?? {}, {});
```
re-up on this question.

hectorhdzg replied:
About meta, this is v2 field so is removed now, related to splat, we need the for loop to clean up undesired values, Object.assign will add a bunch of stuff that is not needed
![image](https://github.com/open-telemetry/opentelemetry-js-contrib/assets/39923391/38971aba-8722-4fd3-9a04-dfcb5d911029)

trentm commented on packages/winston-transport/src/utils.ts:
```diff
@@ -0,0 +1,75 @@
+      attributes[key] = splat[key];
+    }
+  }
+  const logRecord: LogRecord = {
```
Note: This is where instrumentation-bunyan intentionally drops the `trace_id`, `span_id`, and `trace_flags` fields that might have been added by the enabled "log correlation" functionality.  

However, to not hold up this PR, this could be deferred to a separate issue/PR along with the outstanding question -- https://github.com/open-telemetry/opentelemetry-js-contrib/pull/1837#issuecomment-1936798374 -- about the LogRecord `traceId`, `spanId`, and `traceFlags` fields being incorrectly set in cases where there is stream buffering.

hectorhdzg replied:
Created https://github.com/open-telemetry/opentelemetry-js-contrib/issues/2010 to follow up on this

trentm reviewed (changes requested):
I'm really sorry for the block and long delay. I have a couple small changes that I think should be done (e.g. the package-lock.json vestigial entry), a few nits (i.e. feel free to ignore "nits" from me at your discretion), and then other Qs can be deferred to separate issues.

pichlermarc commented on packages/winston-transport/README.md:
```diff
@@ -0,0 +1,70 @@
+[![NPM Published Version][npm-img]][npm-url]
+[![Apache License][license-image]][license-image]
+
+This module provides a Transport for [`winston`](https://www.npmjs.com/package/winston) module to send Winston logging to the OpenTelemetry Logging SDK. It may be loaded using the [`@opentelemetry/sdk-trace-node`](https://github.com/open-telemetry/opentelemetry-js/tree/main/packages/opentelemetry-sdk-trace-node) package.
```
Yes this looks like outdated info to me, probably best to remove it.

hectorhdzg commented on plugins/node/opentelemetry-instrumentation-winston/package.json:
```diff
@@ -41,13 +41,15 @@
   "peerDependencies": {
-    "@opentelemetry/api": "^1.3.0"
+    "@opentelemetry/api": "^1.3.0",
+    "@opentelemetry/api-logs": "^0.49.1"
```
Winston instrumentation doesn't have a dependency to api-logs anymore, the dependency is now in winston transport package, it should be as dependency and not peer dependency like @pichlermarc mentioned. Removing here.

trentm commented:
@hectorhdzg I see you resolved a number of the conversations above, but I don't see any new commits pushed to your feature branch. Did you perhaps forget to push up your local changes? (Sorry if I'm jumping the gun. I'm trying to not be the slow reviewer here again. :)

hectorhdzg commented:
@trentm sorry I usually start resolving whenever I update locally, and push changes after, everything should be up to date now.

trentm reviewed (approved):
LGTM. Thanks!
